src/csharp/SDK/Allocator.cs

```diff
@@ -0,0 +1,411 @@
+        }
+
+        private class BufferCacheEntry
+        {
+            public byte[] ManagedBufferCache { get; set; }
+
+            public int UsedSize { get; set; }
+
+            public int ReferenceCount { get; set; }
+
+            public bool Initialized { get; set; }
+        }
+
+        private void PopulateEntry(IntPtr nativeAddress, BufferCacheEntry entry)
+        {
+            Marshal.Copy(nativeAddress, entry.ManagedBufferCache, 0, entry.UsedSize);
+        }
+
+        /// <summary>
+        /// Get a managed array to cache the contents of a native buffer.
+        /// </summary>
+        /// <param name="nativeAddress">Native buffer to mirror.</param>
+        /// <param name="size">Size of the native memory.</param>
+        /// <returns>A managed array populated with the content of the native buffer.</returns>
+        /// <remarks>Multiple callers asking for the same address will get the same buffer.
+        /// When done with the buffer the caller must call <seealso cref="ReturnBufferCache(IntPtr)"/>.
+        /// </remarks>
+        public byte[] GetBufferCache(IntPtr nativeAddress, int size)
+        {
+            BufferCacheEntry entry;
+
+            lock (this)
+            {
+                if (this.bufferCache.ContainsKey(nativeAddress))
+                {
+                    entry = this.bufferCache[nativeAddress];
+                    entry.ReferenceCount++;
+
+                    if (entry.UsedSize != size)
+                    {
+                        throw new Exception("Multiple image buffers sharing the same address cannot have the same size");
+                    }
+
+                }
+                else
+                {
+                    entry = new BufferCacheEntry
+                    {
+                        ManagedBufferCache = this.pool.Rent(size),
+                        UsedSize = size,
+                        ReferenceCount = 1,
+                        Initialized = false,
+                    };
+
+                    this.bufferCache.Add(nativeAddress, entry);
+                }
+            }
+
+            lock (entry)
+            {
+                if (!entry.Initialized)
+                {
+                    this.PopulateEntry(nativeAddress, entry);
+                    entry.Initialized = true;
+                }
+            }
+
+            return entry.ManagedBufferCache;
+        }
+
+        /// <summary>
+        /// Return the buffer cache.
+        /// </summary>
+        /// <param name="nativeAddress">Address of the native buffer.</param>
+        /// <remarks>Must be called exactly once for each buffer provided by <see cref="GetBufferCache(IntPtr, int)"/>.</remarks>
+        public void ReturnBufferCache(IntPtr nativeAddress)
+        {
+            lock (this)
+            {
+                BufferCacheEntry entry = this.bufferCache[nativeAddress];
+                entry.ReferenceCount--;
```

Comment:
Do you need to take a lock on "entry"? Just to make sure that the initialization isn't still happening? If not you might want to consider using the Interlocked Exchange or Increment methods.

Reply:
No. This this lock ensures that the state of bufferCache is correct. This entry lock is only needed to ensure the memory is initialized.